autoload/cmake/fileapi.vim

```diff
@@ -0,0 +1,188 @@
+        return 0
+    elseif l:index_path ==# s:fileapi.lastIndexName
+        return 2
+    endif
+    let s:fileapi.lastIndexName = l:index_path
+
+    let l:index = json_decode(readfile(l:index_path))
+    let l:reply = l:index.reply['client-' . s:client_name]['query.json']
+    if l:reply.client.query.version != s:query_version
+        " TODO: Regenerate to update queries or bail
+        " bail for now 
+        call s:logger.EchoError('Api query out of date, run :CMakeGenerate')
+        call s:logger.LogError('Api query out of date, run :CMakeGenerate')
+        return 0
+    endif
+    " Update response references
+    let l:responses = {}
+    for l:response in l:reply.responses
+        let l:path = [a:build_dir] + s:reply_path + [l:response.jsonFile]
+        let l:responses[l:response.kind] = s:system.Path(l:path, v:true)
+    endfor
+    let s:fileapi.version = l:reply.client.query.version
+    let s:fileapi.index.responses = l:responses
+    return 1
+endfunction
+
+" Parse codemodel api response
+"
+" Returns:
+"   Number
+"       0: Failed
+"       1: Updated
+"
+function! s:ParseCodemodel(build_dir) abort
+    let l:path_list = [a:build_dir] + s:reply_path + [s:fileapi.index.responses.codemodel]
+    let l:codemodel_path = s:system.Path(l:path_list, v:true)
+    let l:codemodel = json_decode(join(readfile(s:fileapi.index.responses.codemodel)))
+    call map(l:codemodel.configurations[0].targets, {_, val -> val.name})
+    let s:fileapi.codemodel.targets = l:codemodel.configurations[0].targets
+endfunction
+
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+" Public functions
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+" Check if CMake version is supported
+"
+" Params:
+"     cmake_version : Dictionary
+"         major : Number
+"             cmake major version
+"         minor : Number
+"             cmake minor version
+"
+" Returns:
+"   Boolean
+"       Whether passed cmake version is supported by fileapi
+"
+function! s:fileapi.CheckCMakeVersion(cmake_version) abort
+    let l:cmake_version_comparable =
+                \ a:cmake_version.major * 100 + a:cmake_version.minor
+    let s:fileapi.cmake_version_supported = l:cmake_version_comparable >= 314
+    if !s:fileapi.cmake_version_supported
+        let l:warning =
+                    \ 'fileapi: CMake version not supported.'
+                    \ . ' Certain functionality will not work correctly.'
+                    \ . ' (Minimum supported is CMake 3.14)'
+        call s:logger.EchoWarn(l:warning)
+        call s:logger.LogWarn(l:warning)
+    endif
+
+    return s:fileapi.cmake_version_supported
+endfunction
+
+" Reset all fileapi state
+"
+function! s:fileapi.Reset() abort
+    let s:fileapi.version = 0
+    let s:fileapi.lastIndexName = 'unset' " won't compare true
+    let s:fileapi.index = {}
+    let s:fileapi.codemodel = {}
+    let s:fileapi.codemodel.targets = []
```

Comment:
Having `index` here saves me from returning returning it from `ParseIndex` which already has a return value. It could be a script local variable instead of a class variable though. Otherwise you are correct. Having codemodel as a dict is just a thing I did because we might want to add more keys to it, but this would be easy enough to change in the event we did.

Reply:
Missed the `codemodel_file` part. This is the only part used from the index outside the `ParseIndex` function. This once again is a "we may want to add more later". Will change. The entire logic to detect the codemodel path seems weird now. I don't even use the results...